File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int anc=0;
        int n=fruits.size();
        unordered_map<int,int> umap;
        int res =0;
        for(int i=0;i<n;i++){
            umap[fruits[i]]++;
            while(anc<=i && umap.size()>2){
                umap[fruits[anc]]--;
                if(umap[fruits[anc]]==0) umap.erase(fruits[anc]);
                anc++;
            }
            res=max(res,i-anc+1);
        }
        return res;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
from collections import defaultdict 

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        ans,anc=0,0
        freq = defaultdict(int)
        for i in range(len(fruits)): 
            freq[fruits[i]]+=1 
            while anc<=i and len(freq)>2: 
                freq[fruits[anc]]-=1
                if freq[fruits[anc]]==0: 
                    del freq[fruits[anc]]
                anc+=1
            ans = max(ans,i-anc+1)

        return ans 